[1단계 - 3단계] 노랑(노주희) 미션 제출합니다.

yesjuhee opened the pull request:
안녕하세요 찰리 노랑입니다 😄 

이전 미션의 코드를 이어서 방탈출 사용자 예약 기능을 추가적으로 구현했습니다!
이번 단계 미션의 변경 사항 커밋 목록 링크입니다 [링크](https://github.com/woowacourse/spring-roomescape-member/pull/198/files/2b94bbe1df72e8a56f45fab4bbe29f5b55d63858..5b84400d299a974da5bef424d756c3bd8d0f0180)

궁금한 점은 코드에 코멘트로 달아 놓았습니다! 이번 미션 잘 부탁드립니다 🙇‍♀️

<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [ ] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->
원시값을 포장하거나 메서드를 최대한 10줄 이내로 줄이려는 등의 노력을 했습니다.

## 웹 개발 경험 체크
🌱 레벨2 시작 이전에, 웹 애플리케이션 개발 경험이 있나요?
- [ ] 아니오. 전혀 경험이 없습니다.
- [ ] 예. 예제 코드를 따라하거나 클론 코딩을 해봤습니다. (ex. 강의를 보고 따라함)
- [x] 예. 웹 애플리케이션을 만들었지만 로컬에서만 실행해봤습니다.
- [ ] 예. 웹 애플리케이션을 만들고 배포해봤습니다. (ex. 클라우드 혹은 리모트 서버에 배포)

## 베이스 코드 선택 체크
- [x] 이전 미션의 내 코드에서 시작 
- [ ] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

solarsalt committed:
docs: add PR template

yesjuhee committed:
chore: 미션1 코드 마이그레이션

yesjuhee committed:
refactor: API 응답 상태 코드 변경

yesjuhee committed:
docs: 1단계 요구사항 정리

yesjuhee committed:
feat: 예약자명은 최소 1글자, 최대 5글자여야한다

yesjuhee committed:
feat: `HH:mm` 형식으로 시간을 입력해야한다

yesjuhee committed:
feat: 예약 생성 요청 형식 검증한다

yesjuhee committed:
feat: 같은 날짜와 시간에 중복 예약을 할 수 없다

yesjuhee committed:
refactor: Reservation 객체 생성 방식 변경

kjyyjk committed:
feat: 예약 시간을 중복 생성할 수 없다

kjyyjk committed:
feat: 존재하는 시간 ID를 입력해야한다

yesjuhee committed:
feat: 특정 시간에 대한 예약이 존재할 때 예약 시간을 삭제할 수 없다

yesjuhee committed:
feat: 과거에 대한 예약은 불가능하다

kjyyjk committed:
refactor: controller 리팩터링

- 전역적으로 사용되는 ExceptionHandler를 ControllerAdvice로 분리했다.

kjyyjk committed:
refactor: ReservationDateTime 추가

kjyyjk committed:
refactor: Dao 인터페이스 분리

yesjuhee committed:
test: 테스트용 페이크 객체 생성

yesjuhee committed:
test: ReservationService 테스트 작성

kjyyjk committed:
test: ReservationTimeService 테스트 작성

kjyyjk committed:
refactor: 예약자명을 MemberName으로 분리

yesjuhee committed:
docs: 2단계 요구사항 정리

yesjuhee committed:
feat: 테마 관리 페이지 응답 API 추가

yesjuhee committed:
feat: 테마 DDL 추가

kjyyjk committed:
feat: Theme 도메인 객체 추가

yesjuhee committed:
feat: 테마 생성 API 구현

kjyyjk committed:
feat: 테마 생성 API 컨트롤러 구현

kjyyjk committed:
feat: 테마 목록 조회 API 구현

yesjuhee committed:
refactor: 예약 API에 테마 적용

kjyyjk committed:
feat: 테마 삭제 API 구현

yesjuhee committed:
refactor: 컨트롤러 응답 헤더 리팩터링

yesjuhee committed:
chore: 응답 HTML 변경, 경고창 추가

kjyyjk committed:
feat: 사용자 화면 응답 API 추가

yesjuhee committed:
docs: 예약 가능 시간 API 요구사항 추가

yesjuhee committed:
feat: 예약 가능 시간 조회 API 추가

kjyyjk committed:
docs: 인기 테마 조회 API 요구사항 추가

kjyyjk committed:
chore: 목업 데이터 추가

kjyyjk committed:
feat: 인기 테마 조회 API 구현

yesjuhee committed:
feat: 검증 로직 추가

kjyyjk committed:
refactor: Service, Dao 리팩터링

yesjuhee committed:
fix: 생략된 Valid 어노테이션 추가

yesjuhee committed:
chore: 생략된 final 추가

yesjuhee commented on src/main/java/roomescape/controller/ReservationTimeController.java:
```diff
@@ -22,19 +26,32 @@ public ReservationTimeController(final ReservationTimeService reservationTimeSer
+            @RequestParam("themeId") long themeId
+    ) {
+        List<AvailableTimeResponse> responses = reservationTimeService.findAvailableTimes(date, themeId);
+        return ResponseEntity.ok().body(responses);
```
.

yesjuhee commented on src/main/java/roomescape/controller/ReservationTimeController.java:
```diff
@@ -22,19 +26,32 @@ public ReservationTimeController(final ReservationTimeService reservationTimeSer
+    ) {
+        final List<AvailableTimeResponse> responses = reservationTimeService.findAvailableTimes(date, themeId);
+        return ResponseEntity.ok().body(responses);
     }
```
.

yesjuhee replied:
이번 미션의 3단계에서 추가된 예약 가능한 시간 조회 API를 처리하는 컨트롤러입니다.

요청 경로를 /times/available, /available-times 둘 중에서 고민하다가 요청하는 리소스를 명시하기 위해서 명사를 사용하는게 더 적절한다는 결론을 내렸고, 지금과 경로로 결정했습니다.

또한 이 하나의 API를 위해서 AvailableTimesController와 같이 별도의 컨트롤러를 사용하는 것은 불필요하다고 생각해서 ReservationTimeController에 포함시켰습니다.

yesjuhee committed:
chore: 안쓰는 코드 삭제

yesjuhee commented on src/main/java/roomescape/controller/ReservationTimeController.java:
```diff
@@ -0,0 +1,57 @@
+        return ResponseEntity.noContent().build();
+    }
+
+    @GetMapping("/available-times")
+    public ResponseEntity<List<AvailableTimeResponse>> findAvailableTimes(
+            @RequestParam("date") LocalDate date,
+            @RequestParam("themeId") long themeId
+    ) {
+        final List<AvailableTimeResponse> responses = reservationTimeService.findAvailableTimes(date, themeId);
+        return ResponseEntity.ok().body(responses);
+    }
```
이번 미션의 3단계에서 추가된 **예약 가능한 시간 조회 API**를 처리하는 컨트롤러입니다.

요청 경로를 `/times/available`, `/available-times` 둘 중에서 고민하다가 요청하는 리소스를 명시하기 위해서 명사를 사용하는게 더 적절한다는 결론을 내렸고, 지금과 경로로 결정했습니다.

또한 이 하나의 API를 위해서 `AvailableTimesController`와 같이 별도의 컨트롤러를 사용하는 것은 불필요하다고 생각해서 `ReservationTimeController`에 포함시켰습니다.

yesjuhee committed:
chore: 안쓰는 파일 삭제

yesjuhee commented on src/main/java/roomescape/exception/GlobalExceptionHandler.java:
예외 입력 상황에 대해서 정해진 형식으로 404 BadRequest로 응답을 하도록 전역 ExceptionHandler를 만들었습니다!

yesjuhee commented on src/main/java/roomescape/repository/H2ReservationDao.java:
```diff
@@ -0,0 +1,152 @@
+        return jdbcTemplate.query(sql, reservationRowMapper);
+    }
+
+    @Override
+    public Reservation save(final Reservation reservation) {
+        long id = insertReservationAndRetrieveKey(reservation);
+        return new Reservation(id, reservation.getName(), reservation.getDate(), reservation.getTime(),
+                reservation.getTheme());
+    }
```
### DAO의 저장 메서드 작동 방식 질문

DAO에서 save()를 하고 결과값을 반환하는 방식에 대해서 페어와 저의 방식이 달랐는데, 이에 대해서 찰리의 의견을 묻고 싶습니다!

1. 지금처럼 DB에 저장 후 `new Reservation()`으로 `id`를 포함한 객체를 생성해서 반환하는 방식.
2. DB에 저장 후 DAO의 `getReservationById(final long id)`를 호출해서 DB에 저장된 값을 직접 불러오는 방식 

1번으로 할 경우 코드가 간단해지고, 쿼리를 한 번만 날리면 된다는 장점이 있지만, 실제 DB를 통해 가져오는 방식이 아니기 때문에 제대로 저장되었는지 확인이 되지 않을 수도 있다고 생각했습니다.
2번으로 할 경우에는 INSERT, SELECT 2번의 쿼리를 보내야 한다는 단점이 있습니다.

우선은 대부분의 크루들이 1번 방식으로 한 것을 보고 해당 방식을 선택했는데 이게 좋은 선택인가에 대해 확신이 잘 들지 않습니다. 혹시 찰리는 어떻게 생각하시나요?

Gomding replied:
노랑의 말대로 1번은 쿼리 한번으로 간단하게 처리할 수 있다는 장점과 실제 저장된 객체와 동일하지 않을 수 있다는 불안정함
2번은 쿼리는 한 번 더 날리지만 DB 와 완전히 일치한다는 장점이 있겠네요 :)

Reservation을 저장할 때 요청한 값과 다르게 저장될 확률이 존재할까요?
해당 부분은 테스트 코드로 커버할 수도 있겠군요 :)

노랑이 생각하는 `실제 DB를 통해 가져오는 방식이 아니기 때문에 제대로 저장되었는지 확인이 되지 않을 수도 있다고 생각했습니다.` 은 어떤 상황일까요? 🧐

yesjuhee replied:
> Reservation을 저장할 때 요청한 값과 다르게 저장될 확률이 존재할까요? 노랑이 생각하는 실제 DB를 통해 가져오는 방식이 아니기 때문에 제대로 저장되었는지 확인이 되지 않을 수도 있다고 생각했습니다. 은 어떤 상황일까요? 🧐

`insertReservationTimeAndRetrieveKey()` **메서드가 잘못 작성되어서 저장에 실패하는 경우**가 있을 수 있습니다. 

지금 구조에서는 거의 발생하기가 어렵기는 합니다! 만약의 상황을 가정해보자면, 저장하는 메서드에서 데이터의 저장할 칼럼(`start_at`)의 이름을 잘못 표기해서 `null`값이 저장될 수 있을 것 같습니다. 하지만 반환하는 값은 `new Reservation(id, reservationTime.getStartAt())` 이기 때문에 **실제 데이터에는 `null`이 저장되어 있지만 정상적으로 저장된 척이 가능**하다는 점에서 불안정하다고 생각했습니다.

다시 생각해봤을 때 불안정한 것은 맞지만, 트레이드오프 측면에서 이 불안함은 테스트코드를 통해 확인하는 것이 더 적절한 것 같다라는 생각이 드네요ㅎㅎ 어느정도 생각이 정리된 것 같습니다 감사합니다 👍

Gomding replied:
노랑이 설계한 Reservation 객체에는 null이 들어갈 수 없는 구조이기 때문에
만약에 DB 에서 null을 주더라도 예외가 발생해서 비정상 동작할 일은 없겠네요 :)
안전한 상황이라 생각합니다!

yesjuhee commented on src/main/java/roomescape/service/ReservationService.java:
```diff
@@ -1,34 +1,66 @@
+        final ReservationTime reservationTime = findReservationTime(reservationRequest.timeId());
+        final Theme theme = findTheme(reservationRequest.themeId());
+        final Reservation reservation = reservationRequest.convertToReservation(reservationTime, theme);
+        if (reservation.isBefore(LocalDateTime.now())) {
+            throw new IllegalArgumentException("지나간 날짜와 시간은 예약 불가합니다.");
+        }
```
### 검증의 위치

기존 코드에서 검증 로직을 추가할 때 서비스에서의 위치와 도메인 객체에서의 위치에 대해서 페어와 많은 고민을 했습니다. 또한 서비스의 역할과, 도메인 객체가 무엇인가에 대해서도 계속해서 고민하고 있습니다.

제가 지금까지 정리한 것은 아래와 같습니다
- **서비스의 역할**: UI, 데이터 저장과 분리되어서 사용될 수 있는 비즈니스 로직 계층. 비즈니스의 흐름을 관장하고 도메인 객체가 외부 의존성을 가지지 않도록 하는 역할.
- **도메인 객체**: 서비스 계층 안에서 객체끼리의 협력으로 비즈니스 로직을 처리하는 역할.

검증에 대한 비즈니스 요구사항 중, 객체 스스로가 처리할 수 있는 것(ex. 예약자의 이름 길이)은 **도메인 객체 내부에서 처리**하도록 하고, 외부 의존성을 가지는 것(ex. 기존에 존재하는 예약인지 DB를 통해 확인)은 **서비스 메서드에서 처리**했습니다.

### 질문

그런데 표시한 부분처럼 **예약 시간이 과거인지 확인하는 로직**의 경우 서비스에서 처리할지, 도메인 객체 내부에서 처리할지 애매한 경계에 있는 것처럼 느껴져서 찰리의 의견을 여쭤보고 싶습니다.

애매한 경계라고 느낀 이유는 아래와 같습니다

해당 검증 로직은 외부 의존성이 없습니다. 그래서 `Reservation::validateIsAfter`와 같이 메서드를 만들어서 해당 예약이 가지는 시간이 기준 날짜보다 미래의 시간이 아닐 경우 예외를 터트리도록 만들 수 있습니다.

하지만 "예약이 과거의 시간이면 안됨"이라는 조건은, `Reservation` 도메인 자체의 조건이 아니라, 예약을 생성하는 로직의 하위 조건이라는 점에서 `ReservationService::createReservation`이 예외를 발생시키는 적절한 위치라고 생각했습니다.

그러면서 또 한편으로는 `Reservation` 객체에게 메시지를 보내서 스스로 판단하도록 할 수 있는 일을 절차지향적으로 처리하고 있는 코드가 아닌가 하는 생각이 들었습니다.

이를 결정할 기준을 정하지 못해서, 찰리의 추가 의견을 듣고 싶습니다 😄

Gomding replied:
아래 질문으로 의견을 나눠보면 좋겠어요!

- `새로운 예약이 현재 시간보다 과거의 시간을 가진다` 해당 상태를 가진 신규 예약 객체가 생성될 수 있을까요?
- 다른 Service 에서 새로운 예약 객체를 생성할 때 해당 검증을 깜빡한다면 어떤 상황이 발생할까요! 😃 

> 하지만 "예약이 과거의 시간이면 안됨"이라는 조건은, Reservation 도메인 자체의 조건이 아니라, 예약을 생성하는 로직의 하위 조건이라는 점에서 ReservationService::createReservation이 예외를 발생시키는 적절한 위치라고 생각했습니다.

도메인 자체의 조건이 아니라는 이유가 궁금해요!

yesjuhee replied:
> `새로운 예약이 현재 시간보다 과거의 시간을 가진다` 해당 상태를 가진 신규 예약 객체가 생성될 수 있을까요?

넵, 좀 더 정확히 말씀을 드리자면 아래와 같습니다!
`POST /reservations` API를 통해 생성되는 예약 객체의 경우에는 현재 시간보다 과거의 시간을 가질 수 없습니다.
하지만 `new Reservation(...)`을 통해서 새로운 예약 객체를 생성할 때는 현재 시간보다 과거의 시간을 가진 예약 객체가 생성될 수 있습니다.

예를들어 최초 예약을 생성하고, 시간이 지나서 `GET /reservations`로 예약 목록을 불러올 때 해당 목록에 과거의 예약이 포함되어 있으면 `new Reservation()`을 통해서 예약 객체를 생성할 때 과거의 시간을 가진 예약 객체가 생성될 수 있다고 생각합니다.

그래서 **"예약이 과거의 시간이면 안됨"이라는 조건은, Reservation 도메인 자체의 조건이 아니라, 예약을 생성하는 로직의 하위 조건**이라고 판단했습니다 😄

Gomding replied:
예약을 생성하는 로직이라는게 Reservation 객체 생성 규칙을 의미하는게 아닐까요? 🤔 
비즈니스 흐름인지 도메인 규칙인지 궁금합니다!

Service 는 Reservation 객체 생성을 호출하고 
Reservation 내부에는 `새로 생성하는 예약이 현재보다 과거 시간을 가지면 생성할 수 없다` 규칙이 있는것이 어색하지 않다고 생각해요 :)
물론 과거 예약객체 생성도 필요하니 과거 예약 객체를 생성하는 통로도 제공해야겠죠

제 의견은 `도메인 규칙을 Service 계층에 노출해야할까?` 라는 부분입니다!
새로운 예약 생성이라는 목적을 위해 Reservation 객체를 생성한다면
ReservationService 뿐만 아니라 모든 곳에서 해당 규칙이 적용돼야한다고 생각해요
그 규칙을 공통으로 적용하기 좋은곳이 Reservation 객체이구요

yesjuhee replied:
> 예약을 생성하는 로직이라는게 Reservation 객체 생성 규칙을 의미하는게 아닐까요? 🤔

다시 곰곰이 생각해보니 제가 잘못 생각하고 있었던 것 같습니다. "객체 생성 규칙"을 단순히 **new 연산자를 써서 새로운 인스턴스를 생성하는 것**으로 단순히 이해하고 있었네요. 생성자에 넣어줄 만 한 규칙이 아니어서 도메인 자체의 조건이 아니라고 착각하고 있었습니다!

"지나간 날짜와 시간은 예약할 수 없음"은 도메인 규칙이기 때문에 서비스보단 도메인에 위치시키는게 적절한 것 같습니다. 대신 정적팩토리메서드를 사용해 상황에 맞게 이 규칙을 적용한 인스턴스를 생성하도록 변경해보겠습니다!

yesjuhee replied:
e5ca1feb

처음엔 정팩메 도입을 고려했었는데, 최초 예약이 생성되는 시점은 "id가 없는 예약이 생성되는 시점"으로 이해하면 정팩메가 없어도 될 것 같습니다. id 값을 받지 않는 생성자에서만 과거 시간이 아닌지 검증하는 로직을 추가했습니다!

yesjuhee committed:
chore: 안쓰는 파일 삭제

Gomding commented on src/main/java/roomescape/controller/ReservationTimeController.java:
```diff
@@ -0,0 +1,57 @@
+    @GetMapping("/available-times")
+    public ResponseEntity<List<AvailableTimeResponse>> findAvailableTimes(
+            @RequestParam("date") LocalDate date,
+            @RequestParam("themeId") long themeId
```
Long(Wrapper) 타입을 사용할 수도 있는데 long 타입을 선택하신 배경이 궁금해요!

yesjuhee replied:
래퍼 타입의 경우 `null`이 들어올 수 있다는 점 때문에 원시타입(`long`)보다 안정성이 떨어진다고 생각합니다.
그래서 꼭 래퍼타입이 필요한 경우(nullable 한 변수)가 아니면 `long`을 우선적으로 사용하고 있습니다!

Gomding replied:
0 이라는 ID는 실제로 존재할 수 있지 않을까요? 🤔 
ID 가 없음을 표현하는데 충분할까 라는 고민이 있습니다 😃 
노랑은 어떻게 생각하시나요?

yesjuhee replied:
<img width="658" alt="image" src="https://github.com/user-attachments/assets/8601b0db-d439-4e08-91d5-ec1456596568" />

기록을 위해 남깁니다!

yesjuhee replied:
cb6ac104

Gomding replied:
또 애매한 코멘트 있다면 언제든 질문주세요!! 😔

Gomding commented on src/main/java/roomescape/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,71 @@
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+
+@ControllerAdvice
+public class GlobalExceptionHandler {
```
`@ControllerAdvice` 를 사용한 글로벌 예외 핸들러 사용 👍👍👍

Gomding commented on src/main/java/roomescape/exception/ErrorResponse.java:
```diff
@@ -0,0 +1,6 @@
+
+import java.time.LocalDateTime;
+
+public record ErrorResponse(LocalDateTime timestamp, int status, String error, String message, String path) {
```
에러 응답에 사용할 객체를 구현해주셨네요 👍 👍 👍 
클라이언트에게 일관된 형태로 예외응답을 줄 수 있겠어요 :)

몇가지 의문이 있어요!
- error 와 message 는 어떤 값을 넣어줘야할까요?
- path 가 필요한 이유는 무엇인가요?
- HTTP 응답에 status code 가 있는데 status 가 필요한 이유는 무엇인가요?
- timestamp 는 어떤 의도로 응답에 주는것일까요?

yesjuhee replied:
```
{
    "timestamp": "2025-05-02T21:32:08.668368", # 에러 발생 시점
    "status": 400, # 에러 코드                                 
    "error": "Bad Request", # 응답 구문
    "message": "시간을 입력해주세요.", # 에러 세부사항 설명
    "path": "/times" # 요청 경로
}
```

위의 형식으로 값을 보내주고 있습니다! 
페어를 할 때 단순히 메시지만 보내주는 것 보다는 좀 더 친절한 응답형식을 지정해서 사용해보자라는 이야기를 나눴습니다. 그런데 당시에는 시간이 없기도 했고, 적절한 예외 응답에 대한 기준이 둘다 없었기 때문에 **구글링해서 많이 사용하는 방식**을 선택해서 사용했습니다ㅎㅎ

우선 질문해주신 것에 기반해서 생각을 해봤을 때 아래와 같이 생각할 수 있을 것 같습니다.
- `timestamp`는 개발 과정에서 디버깅을 할 때 에러 추적에 유용하게 사용할 수 있을 것 같습니다. 다만 꼭 응답에 포함되어야할까? 라는 생각을 했습니다.
- `status`, `message`, `path`와 같은 값들을 body에 넣어서 보내주면 중복 정보이지만 클라이언트 개발자가 응답 바디만 보고 문제 상황을 더 빠르게, 편하게 파악할 수 있기 때문에 친절하게 정보를 제공하기 위해 body에 포함할 수 있을 것 같습니다. 또 한편으로는 해당 정보들은 이미 HTTP 규약 속에 정해진 위치가 있는데, 친절하게 제공하는게 의미가 있나하는 생각도 드네요!

이 부분에 대해서는 추가 학습 후, 좋은 응답 형식에 대한 기준을 고민해보고 리팩터링하도록 하겠습니다!

Gomding replied:
> 좀 더 친절한 응답형식을 지정해서 사용해보자라는 이야기를 나눴습니다.

훌륭합니다 ㅎㅎㅎ 💯 💯 💯 💯 

`timestamp` 는 실제로 서버에서 응답한 시간과 클라이언트가 응답을 받은 시간을 볼 수 있어서 좋다고 생각해요 :)
`status, message, path` 는 꼭 필요하지않다면 응답객체를 생성할 때 귀찮음이 발생하지 않을까 생각했어요 😃

yesjuhee replied:
cb6ac104

예외 처리 형식에 대해서는 STEP2 하면서 조금 더 고민해보도록 하겠습니다..!
우선 미처 처리하지 못한 예외 상황에 대한 Handler만 추가했습니다.
2단계에서는 400, 404, 401, 403, 500 등 상황에 맞는 상태 코드를 사용해보려고 합니다.

Gomding commented on src/main/java/roomescape/repository/H2ReservationDao.java:
```diff
@@ -0,0 +1,152 @@
+import roomescape.service.reservation.Theme;
+
+@Repository
+public class H2ReservationDao implements ReservationDao {
```
해당 클래스 내부에 H2 에 대한 내용이 있을까요? 😃

yesjuhee replied:
`ReservationDao`인터페이스의 구현체를 만들 때, 그 구현체는 사용하는 저장소의 종류에 따라 달라질 수 있다고 생각해서 구현체의 이름에 H2를 명시했습니다! **해당 클래스에서 사용하는 SQL 문법이 H2에 대한 내용**이라고 볼 수 있을 것 같습니다.

만약 MySQL 저장소가 추가된다면에 대한 생각도 해봤습니다. DBMS 종류마다 방언이 존재하는 것으로 알고 있어서 결국 `MySqlReservationDao`를 생성하게 되지 않을까하고 생각했습니다!

Gomding replied:
쿼리와 데이터베이스 엔진 관점에서 추상화를 해주셨군요 😃 
다른 관점에서도 의견을 나눠보죠!

그렇다면 변경될 부분은 SQL 쿼리 문자열뿐일 것 같은데 쿼리 문자열만 교체해주면 문제가 없다고 생각해요 :)
즉 데이터베이스 벤더를 변경하더라도 쿼리 문자열만 변경될것으로 예상됩니다!
문자열만 변경하는거라면 상수로 관리해도 쉽게 변경할 수 있겠네요 :)

데이터베이스 접근을 위한 JdbcTemplate 측면에서 고민을 해보면
데이터 접근 기술이 변경되면 H2ReservationDao 는 어떻게 될까요?
JdbcTemplate 말고 대표적으로 JPA 가 있는데 JdbcTemplate 에서 JPA 로 넘어가는 상황을 상상해볼 수 있곘네요 :)

+) 특정 DB 벤더에서만 작동하는 SQL 문법을 마구 사용한다면
미래에 DB 벤더 변경이 정말 필요할 때 많은 리팩터링이 필요할 수 있구요
또 실서버에서는 MYSQL 을 사용하는데 테스트환경에서는 H2 를 쓴다면 

MySQL 에 특화된 쿼리를 사용한다면 H2를 사용하는 테스트 환경에서 곤란해 질 수 있어요
테스트에서는 MySQL 에서 특화된 쿼리를 어떻게 실행 해야하나 라는 고민도 생길 수 있죠

yesjuhee replied:
71ed40d7

아하 해당 DAO를 H2 DAO의 관점으로 보는 것은 너무 한정적인 용도의 DAO로 보게되는 문제가 있군요
`JdbcThemeDao`와 같은 네이밍으로 전부 변경했습니다!

Gomding commented on src/main/java/roomescape/repository/H2ThemeDao.java:
```diff
@@ -0,0 +1,90 @@
+
+    @Override
+    public boolean isExists(final ThemeName name) {
+        final String sql = "SELECT COUNT(*) FROM theme WHERE name = ?";
```
존재 여부만을 판단하기위해 query를 실행하는데
전체 개수를 카운트할 필요가 있을까요!

다른 적절한 쿼리는 없을지 확인해보면 좋겠어요 :)
- https://jojoldu.tistory.com/516

yesjuhee replied:
1b2ee45a

안그래도 더 효율적인 방법이 없을까 생각했는데 새로운 방법 배워갑니다!

yesjuhee replied:
1b2ee45a

안그래도 더 효율적인 방법이 없을까 생각했는데 새로운 방법 배워갑니다!

Gomding commented on src/main/java/roomescape/service/reservation/Reservation.java:
```diff
@@ -0,0 +1,74 @@
+package roomescape.service.reservation;
```
`Reservation` 클래스가 service 패키지의 하위에 있군요
노랑이 생각하는 service 패키지는 어떤 클래스를 모아둔 곳인가요?
`Reservation` 객체는 어떤 책임을 가지고있나요?

yesjuhee replied:
좋은 질문 감사합니다 👍 

`service` 패키지는 서비스 계층과 관련이 있는 클래스들을 모아놓은 패키지입니다. `Reservation`과 같은 도메인 객체가 서비스 계층을 구성한다고 생각했습니다!

위에서 언급했듯이 제가 지금까지 정리한 서비스와 도메인은 다음과 같습니다.
- 서비스의 역할: UI, 데이터 저장과 분리되어서 사용될 수 있는 비즈니스 로직 계층. 비즈니스의 흐름을 관장하고 도메인 객체가 외부 의존성을 가지지 않도록 하는 역할.
- 도메인 객체: 서비스 계층 안에서 객체끼리의 협력으로 비즈니스 로직을 처리하는 역할.

제가 정의한 `Reservation` 객체는 서비스에서 비즈니스 로직을 처리할 때, "예약"에 대한 책임을 가지고 다른 객체와 협력하는 도메인 객체입니다!

Gomding replied:
그렇다면 서비스 계층이 도메인을 포함했다고 봐야할까요?
도메인 계층이 서비스에 종속된 계층인지 고민해보면 좋겠어요!

![image](https://github.com/user-attachments/assets/1aac45d9-b4c0-4e43-afdc-f7e188f9e743)

yesjuhee replied:
353e5307

레이어드 아키텍처를 공부하면서 오개념을 가지고 있었네요!
Controller-Service-Repository 3개의 계층으로 구성이 되어있다고 해서 다른 구성요소들이 모두 그 계층에 속할 필요는 없다는 것을 배웠습니다.

한가지 추가 질문이 있습니다! 지금 패키지 구조에서, 위의 그림에 맞춰 Web-Service-Repository 계층으로 수정해주는게 좋을까요?

현재 구조
```
src/main/java/roomescape
├── RoomescapeApplication.java
├── controller
├── domain
├── dto
├── exception
├── repository
└── service
```

그림에 맞춘 구조
```
src/main/java/roomescape
├── RoomescapeApplication.java
├── web
│   ├── controller
│   └── exception
├── domain
├── dto
├── repository
└── service
```

Gomding replied:
정하기 나름일 것 같지만
web 이나 api 등의 이름으로 많이 짓는것 같아요!
말씀해주신대로 web 내부에 exception과 controller 를 구현해주시는것도 좋아요~

exception 내부를 봤을 때는 예외 응답 처리를 위한 내용만 담겨있다고 생각해서요!
만약 현재처럼 doamin, repository, service 와 같은 레벨에 존재한다면
커스텀 예외가 담겨있는 패키지라고 착각할 것 같아요

Gomding commented on src/main/java/roomescape/service/reservation/ThemeName.java:
```diff
@@ -0,0 +1,41 @@
+    @Override
+    public int hashCode() {
+        return Objects.hashCode(name);
+    }
```
꼼꼼한 equals , hashCode 재정의 👍👍👍

Gomding commented on src/main/java/roomescape/service/reservation/MemberName.java:
```diff
@@ -0,0 +1,21 @@
+package roomescape.service.reservation;
+
+public final class MemberName {
```
이름값도 포장해주셨군요 💯 💯 💯

yesjuhee replied:
감사합니다ㅎㅎ
좀 더 설명을 덧붙이자면 서비스가 외부 의존성 처리하도록 해서 도메인 객체는 외부 의존성이 없는 순수 객체끼리 소통하게 설계했습니다. 이 설계를 바탕으로 도메인 객체들은 레벨1에서 배웠던 OOP를 최대한 적용해보려고 했고 이름도 포장했습니다!

Gomding commented on src/test/java/roomescape/service/reservation/MemberNameTest.java:
```diff
@@ -0,0 +1,23 @@
+    @ParameterizedTest
+    @ValueSource(strings = {"aaaaaa", " ", "   "})
+    @NullAndEmptySource
+    void testValidateName(String name) {
```
null or empty 테스트와 이름 길이에 대한 테스트를 분리해보는것은 어떻게 생각하시나요?

하나의 테스트에 여러 조건이 포함돼서
이름에 어떤 조건이 있는지 빠른 파악이 조금 힘든것 같네요 🤔

Gomding commented on src/test/java/roomescape/repository/ReservationDaoTest.java:
```diff
@@ -0,0 +1,116 @@
+    private DataSource dataSource;
+
+    @BeforeEach
+    void setUp() {
```
sql 파일과 `@Sql` 애너테이션을 사용해서 해당 부분을 구현해볼 수 도 있겠어요!

yesjuhee replied:
ecd68e19

DAO 테스트를 전체적으로 추가하면서 해당 내용 반영했습니다 🫡

Gomding commented on src/test/java/roomescape/repository/ReservationDaoTest.java:
```diff
@@ -0,0 +1,116 @@
+import roomescape.service.reservation.Theme;
+
+@JdbcTest
+class ReservationDaoTest {
```
다른 Dao 클래스에 대한 테스트는 작성하지 않아도 괜찮을까요?
노랑의 테스트 작성 기준이 궁금해요! 😃

yesjuhee replied:
작성이 필요하다고 생각합니다...ㅎㅎ
시간 관계상 모든 테스트를 작성하지 못했고, 이번 미션에서는 최우선적으로 서비스 도메인 단위테스트를 선택해서 작성했습니다!
TDD 사이클을 이용해서 개발했기 때문에 JUnit만 사용해서 빠르게 테스트를 작성하고 실행할 수 있는 테스트가 필요했고, 또한 테스트하고 싶은 대부분의 로직이 서비스코드에 집중되어 있었기 때문에 서비스 도메인 단위테스트를 최우선적으로 작성했습니다.

제가 짠 DAO를 안심하고 사용하기 위해서는 DAO 테스트가 모두 있어야 한다고 생각합니다. 또한 클라이언트 개발자가 사용하는 단위의 기능을 검증하기 위해서 통합테스트도 추가해야할 것 같습니다!

추가 작성 후 리뷰요청 드리겠습니다 🫡

yesjuhee replied:
ecd68e19

DAO 테스트 추가했습니다!
RestAssured를 사용한 서버 단위의 E2E 테스트도 추가하고 싶었는데, step1이 좀 늦어지는 것 같아서 해당 테스트는 step2 구현하면서 추가하겠습니당

Gomding reviewed (changes requested):
안녕하세요 노랑!
방탈출 사용자 예약 미션 함께하게된 찰리입니다 :)
미션 꼼꼼하게 잘 진행해주셨네요 👍 👍 
몇가지 코멘트 남겼으니 확인부탁드려요!

궁금한 점 있으면 편하게 DM이나 코멘트 남겨주세요~

yesjuhee commented:
정성스러운 리뷰 감사합니다 🫶 
모든 리뷰 확인하고, 의견이 필요한 리뷰에는 답글을 남겼습니다!

아직 리뷰를 반영해서 코드를 수정하지는 않았는데요, 
대부분의 리뷰가 코드 반영보다는 찰리와 대화가 필요한 부분인 것 같아서 우선 찰리의 질문에 답변을 남겨놓은 것들을 빨리 이야기 하고 싶어서 리뷰 요청을 보냈습니다! 코드 반영도 최대한 빨리 해보겠습니다 🫡

Gomding commented on src/main/java/roomescape/service/ReservationTimeService.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    private boolean isAlreadyBooked(final ReservationTime reservationTime, final List<Reservation> reservations) {
+        for (final Reservation reservation : reservations) {
+            if (reservation.isSameTime(reservationTime)) {
+                return true;
+            }
+        }
+        return false;
```
depth 가 2로 들어갔군요 :)
레벨1과 동일한 규칙을 유지해야하니 depth1 을 유지할 수 있게 리팩터링 해보죠!

yesjuhee replied:
27987903

반영했습니다!

yesjuhee replied:
fad8900 관련 API를 리팩터링하면서 해당 메서드는 삭제됐습니다

Gomding commented on src/main/java/roomescape/service/ReservationTimeService.java:
```diff
@@ -0,0 +1,64 @@
+        reservationTimeDao.deleteById(id);
+    }
+
+    public List<AvailableTimeResponse> findAvailableTimes(final LocalDate date, final long themeId) {
+        final List<ReservationTime> reservationTimes = reservationTimeDao.findAll();
+        final List<Reservation> reservations = reservationDao.findAllByDateAndThemeId(date, themeId);
+        return reservationTimes.stream()
+                .map(time -> new AvailableTimeResponse(time.getId(), time.getStartAt(),
+                        isAlreadyBooked(time, reservations)))
+                .toList();
+    }
```
현재는 예약 가능한 시간을 조회하기위해 쿼리 두 번을 실행하고 있는데
한 번의 쿼리로 실행할 수 있는 방법을 없을까요!

yesjuhee replied:
fad89006

좋은 리뷰 감사합니다 👍 

한 번의 쿼리로 예약 가능 정보를 불러오도록 수정하면서 **`ReservationAvailability` 도메인을 추가**했습니다. 

또한 해당 도메인에 맞게 **`ReservationAvailabilityRpository`를 추가**했습니다. 지금까지는 DB 테이블과 DAO를 1:1 관점으로 생각했었는데 해당 기능을 리팩터링하면서 여기에 갖혀 있으면 안된다는 것을 느꼈습니다. 이전 미션을 하면서 "DAO는 DB에 초점이 맞춰져 있다면, Repository는 도메인 객체에 초점이 맞춰져 있는 것의 차이가 있다.", "Repository는 DB의 저장 형태에 관계 없이 도메인 객체의 저장소 역할을 한다."와 같은 말을 들었을 때 잘 체감이 되지 않았었습니다. 이번에 `ReservationAvailability` 도메인을 정의하면서 레포지터리가 DB의 저장 형태와 관계 없이 도메인 객체의 저장소 역할을 할 때 유용함을 느꼈습니다. 이후 커밋(404873a0)에서 DAO 인터페이스의 이름을 `~Repository`로 수정했습니다.

서비스는 **`ReservationAvailabilityService`를 새로 만들었고**, 컨트롤러는 **`ReservationTimeController`를 재사용**했습니다. 

서비스의 경우 `ReservationAvailablity`도메인에 초점을 맞춰서 새로 생성했습니다. 

컨트롤러는 도메인보단 클라이언트와 소통에 초점을 맞춰 `ReservationTimeController`를 재사용했습니다. 클라이언트가 해당 API를 호출하는 흐름을 고려했을 때, "날짜"와 "테마"를 선택한 후 해당 범위에서 "시간들의 사용 가능 여부 정보"를 GET 요청하는 흐름입니다. 즉 `ReservationAvailability` 도메인 전체에 관심이 있는 것은 아니고, 그 중 시간에 관한 정보만 관심이 있기 때문에 `TimeAvailabilityRespone` DTO를 분리해 사용했습니다. 또한 API 요청 형식도 좀 더 이해하기 쉬운 형태로 변경했습니다(1a390c84)

Gomding commented on src/main/java/roomescape/service/ReservationTimeService.java:
```diff
@@ -0,0 +1,64 @@
+        reservationTimeDao.deleteById(id);
+    }
+
+    public List<AvailableTimeResponse> findAvailableTimes(final LocalDate date, final long themeId) {
```
예약 가능한 시간과 이미 예약된 시간 모두를 조회하고 있는데
메서드 이름은 예약 가능한 시간만 조회될 것으로 예상했어요 🤔 
노랑은 어떻게 생각하세요?

yesjuhee replied:
fad8900 

예약 가능한 시간만 조회될 것으로 예상되는 부분 동의합니다!
관련 API를 리팩터링 하면서 네이밍을 전부 수정했습니다 👍

Gomding commented on src/main/java/roomescape/service/ReservationTimeService.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    public ReservationTimeResponse createReservationTime(final ReservationTimeRequest reservationTimeRequest) {
+        if (reservationTimeDao.isExistsByTime(reservationTimeRequest.startAt())) {
+            throw new IllegalArgumentException("이미 존재하는 시간입니다.");
+        }
```
꼼꼼한 검사와 예외처리 👍 👍

Gomding commented on src/main/java/roomescape/controller/dto/ReservationTimeRequest.java:
```diff
@@ -0,0 +1,12 @@
+import java.time.LocalTime;
+import roomescape.service.reservation.ReservationTime;
+
+public record ReservationTimeRequest(@NotNull(message = "시간을 입력해주세요.") LocalTime startAt) {
+
+    public ReservationTime converToReservationTime() {
+        return new ReservationTime(this.startAt);
+    }
```
요건 유익하군요 👍 👍

yesjuhee replied:
7ae04bd0

Gomding commented on src/main/java/roomescape/service/ThemeService.java:
```diff
@@ -0,0 +1,54 @@
+        themeDao.deleteById(id);
+    }
+
+    public List<ThemeResponse> findPopularThemes() {
```
해당 테스트는 작성하지 못하신것 같네요 😢 

```java
    /**
     * TODO
     * JOIN을 구현할 방법 찾기
     */
    @Override
    public List<Theme> findPopularThemes(LocalDate from, LocalDate to, int count) {
        return null;
    }
```

다음 요청에는 추가 되면 좋겠어요!

Gomding commented on src/test/java/roomescape/fake/FakeThemeDao.java:
```diff
@@ -0,0 +1,61 @@
+
+    @Override
+    public Theme findById(final long id) {
+        return themes.stream()
+                .filter(theme -> theme.getId() == id)
+                .findAny()
+                .orElseThrow();
```
Dao 와 던지는 예외가 다를것 같네요 🤔 
Fake 를 올바르게 구현한게 맞는지 의심해보면 좋겠어요

Gomding commented on src/main/java/roomescape/service/ReservationService.java:
```diff
@@ -0,0 +1,66 @@
+        return reservationTimeDao.findById(timeId);
+    }
+
+    private Theme findTheme(final long themeId) {
+        if (themeDao.isNotExistsById(themeId)) {
+            throw new IllegalArgumentException("테마가 존재하지 않습니다.");
+        }
+        return themeDao.findById(themeId);
+    }
```
ThemeDao 의 findById 가 Optional 을 반환하는것은 어떠신가요!
쿼리를 한번으로 줄이는 효과가 있습니다 😃 

findBtId 호출 시 값이 없을수도 있다는것을 명확하게 표현하여 
값이 없을 때 처리를 강제할 수 있구요

+) 만약 변경하신다면 비슷하게 구현된 부분도 함께 변경해주세요 🙏

yesjuhee replied:
f2b5193f

반영했습니다! `Optional` 활용 방법에 잘 인식하지 못하고 있었는데 많이 배웠습니다 👍

Gomding commented on src/main/java/roomescape/repository/H2ReservationTimeDao.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    @Override
+    public ReservationTime save(final ReservationTime reservationTime) {
+        final long id = insertReservationTimeAndRetrieveKey(reservationTime);
+        return new ReservationTime(id, reservationTime.getStartAt());
+    }
```
`insertReservationTimeAndRetrieveKey` 필요한 메서드 추출인지 의문이 들었어요 😃 
가독성이 오히려 나빠지지 않았는지 고민해보면 좋겠어요!
어떤 장점이 있어서 분리를 해주셨을까요?

yesjuhee replied:
이전에 `save()` 메서드에서 저장 후 아이디로 직접 조회를 해서 반환하는 방식을 사용할 때 저장하는 메서드와 조회하는 메서드를 분리해서 가독성을 높이기 위해 사용했습니다! 아이디로 직접 조회하는 과정이 필요 없어진 지금 불필요한 메서드 분리가 맞는 것 같습니다.
리뷰 반영했습니다!

aed4e408

Gomding reviewed (changes requested):
안녕하세요 노랑!
꼼꼼하게 답변해주셨네요 👍 👍 
추가로 코멘트 남겼으니 확인부탁드려요~!

궁금한 점 있으면 언제든 DM 이나 코멘트 남겨주세요 🙏

yesjuhee committed:
refactor: `findById` 메서드가 `Optional`을 반환하도록 리팩터링

yesjuhee commented on src/main/java/roomescape/repository/H2ThemeDao.java:
```diff
@@ -0,0 +1,90 @@
+    public Theme findById(final long id) {
+        final String sql = "SELECT id, name, description, thumbnail FROM theme WHERE id = ?";
+        return jdbcTemplate.queryForObject(sql, themeMapper, id);
+    }
```
f2b5193f

queryForObject 대신 query 메서드 사용 후 Optional을 반환하도록 처리함

yesjuhee committed:
refactor: 불필요한 메서드 분리 제거, 사용하지 않는 메서드 제거

yesjuhee commented on src/main/java/roomescape/repository/H2ReservationDao.java:
```diff
@@ -0,0 +1,152 @@
+                WHERE r.id = ?
+                """;
+        return jdbcTemplate.queryForObject(sql, reservationRowMapper, id);
+    }
```
aed4e408

반영 완료

yesjuhee committed:
chore: DAO 메서드 순서 정렬

- CRD 순으로 정렬
- 사용하지 않는 메서드 제거

yesjuhee committed:
fix: 잘못된 sql 문법 수정

yesjuhee committed:
refactor: 존재 여부 확인 쿼리의 성능 개선

yesjuhee committed:
chore: DAO 클래스 이름 접두사를 `H2`에서 `Jdbc`로 변경

yesjuhee commented on src/main/java/roomescape/repository/H2ReservationDao.java:
```diff
@@ -0,0 +1,137 @@
+    public boolean isExistsByDateAndTimeId(final LocalDate date, final long timeId) {
+        final String sql = "SELECT count(*) FROM reservation WHERE date = ? AND time_id = ?";
+        Long count = jdbcTemplate.queryForObject(sql, Long.class, date, timeId);
+        return count > 0;
```
1b2ee45a

성능 개선 과정에서 `queryForObject()`를 `query()`로 변경하면서 해당 가능성이 제거됨. `jdbcTemplate.queryForObject()`에서 `NullPointException`의 발생 가능성을 배웠다 토끼 땡큐~

yesjuhee commented on src/main/java/roomescape/repository/H2ReservationTimeDao.java:
```diff
@@ -0,0 +1,77 @@
+    public ReservationTime findById(final long id) {
+        final String sql = "SELECT id, start_at FROM reservation_time WHERE id = ?";
+        return jdbcTemplate.queryForObject(sql, reservationTimeRowMapper, id);
+    }
```
f2b5193f

좋은 포인트야~ `jdbcTemplate.queryForObject()`를 쓸 때는 `NPE` 뿐만아니라 `EmptyResultDataAccessException`도 문제가 되는구나~

yesjuhee committed:
refactor: 메서드 이름 변경, 용도에 맞는 생성자 추가, 값 객체 추가

- `isExists*()`를 `exists*()`로 변경
- id 값을 받지 않는 도메인 객체 생성자 추가
- ThemeThumbnail 값 객체 추가

yesjuhee committed:
test: DAO 테스트 추가

yesjuhee committed:
chore: 사소한 오류 수정

yesjuhee commented on src/test/java/roomescape/repository/JdbcReservationDaoTest.java:
```diff
@@ -0,0 +1,116 @@
+        List<Reservation> actual = jdbcReservationDao.findAllByDateAndThemeId(date, timeId);
+        // then
+        assertThat(actual).hasSize(2);
+    }
```
dbb37e9

토끼야 고마워

yesjuhee committed:
fix: 예약 중복 기준 변경

기존에는 날짜+시간만 확인했는데, 날짜+시간+테마 모두 같아야 중복으로 판단하도록 수정함

yesjuhee committed:
refactor: depth2를 1로 수정

yesjuhee committed:
refactor: 이용가능시간 조회 API 리팩터링

- `ReservationAvailability` 도메인 추가
- `ReservationAvailabilityDao` 추가 및 쿼리 한 번으로 조회 가능하도록 리팩터링
- `ReservationAvailabilityService` 추가
- 요청 경로 변경 (GET /available-times -> GET /times/availability)

yesjuhee committed:
chore: 오타 수정

yesjuhee commented on src/test/java/roomescape/fake/FakeReservationDao.java:
```diff
@@ -0,0 +1,68 @@
+                .anyMatch(reservation ->
+                        reservation.getTheme().getId() == themeId
+                                && reservation.getDate().isEqual(date)
+                                && reservation.getTime().getId() == themeId);
```
7ae04bd0

yesjuhee committed:
refactor: 패키지 구조 수정

yesjuhee committed:
chore: DAO 인터페이스의 이름을 `**Repository`로 변경

yesjuhee committed:
docs: 이용가능시간 조회 API 변경 사항 반영

yesjuhee committed:
refactor: `Reservation` 객체 생성 검증 방식 변경

- 최초 생성시 과거 시간이 아닌지 객체 스스로 검사하도록 리팩터링
- 주생성자에서 null 검증 추가

yesjuhee committed:
test: 잘못 작성된 테스트 수정

yesjuhee committed:
feat: path valiable 예외 처리 추가

yesjuhee commented:
아직 모든 리뷰를 반영하지는 못했지만, 이쯤에서 STEP1을 마무리 하고 그 다음 단계로 넘어가기 위해 리뷰 요청 보냅니다 🙇 🙇‍♀️ 
반영한 리뷰는 🚀 이모지로 표시했고, 아직 반영하지 못하거나 더 논의가 필요한 부분은 👀 이모지 달아두었습니다!

테스트 같은 경우 **DAO 테스트**는 모두 추가했습니다.
RestAssured를 사용한 **E2E 테스트**는 2단계를 진행하면서 추가하도록 하겠습니다!
일부 누락된 **서비스 테스트**도 추가하도록 하겠습니다.

서비스 코드의 단위 테스트 대해서 추가로 질문드리고 싶은게 있습니다! 인기 테마 조회나 시간 이용 가능 여부 조회 같이 JOIN을 사용한 복잡한 쿼리를 사용하는 기능이 추가되니까 **Fake 객체의 단점(구현이 복잡해짐, Fake 객체 자체의 오류 발생 가능)을 실감하고 있습니다**. 따라서 서비스 단위 테스트를 위해 Fake 객체의 사용 대신 Mockito 라이브러리를 활용한 Mock의 도입을 고려하고 있습니다. 다만 너무 성급한 외부 라이브러리 도입이 아닐지 약간 우려가 되어서 찰리의 의견을 여쭙고 싶습니다. **Mockito를 도입하기에 적절한 상황에 맞을까요?** 혹은 Fake 객체의 사용만으로 충분할까요?

Gomding commented on README.md:
```diff
@@ -0,0 +1,320 @@
+# 방 탈출 예약 관리
+
+## 기능 요구사항
```
요구사항을 깔끔하게 정리해주셨네요 💯 💯 💯

Gomding commented on src/main/java/roomescape/domain/Reservation.java:
```diff
@@ -0,0 +1,97 @@
+        this(id, new MemberName(name), new ReservationDateTime(date, time), theme);
+    }
+
+    public Reservation(final String name, final LocalDate date,
+                       final ReservationTime time, final Theme theme) {
+        this(null, new MemberName(name), new ReservationDateTime(date, time), theme);
+        validateDateTime();
+    }
```
해당 생성자를 다른 개발자가 볼 때
`validateDateTime 을 왜 여기서만 검사를 하지?` 라는 추가 의문이 생길 것 같아요
정적 팩터리 메서드를 사용했다면 객체 생성에 대한 이름을 부여할 수 있고
거기서 검사하는 이유에 대한 힌트를 얻지않을까요 😃 

노랑은 어떻게 생각하세요?

yesjuhee replied:
좋은 이유인 것 같습니다! 다음 단계에서 리팩터링 해보겠습니당

Gomding commented on src/main/java/roomescape/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,96 @@
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.web.method.annotation.MethodArgumentTypeMismatchException;
+
+@ControllerAdvice
```
혹시 `@RestControllerAdvice` 도 고려해보셨을까요!

Gomding commented on src/main/java/roomescape/dto/ReservationRequest.java:
```diff
@@ -0,0 +1,19 @@
+package roomescape.dto;
```
dto 는 가능하면 특정 계층에 위치해보면 좋겠어요!
dto 는 계층을 명확하게 하는게 좋다고 생각해서요 :)

노랑이 dto 를 controller, domain, service 와 동일한 레벨에 위치한 이유를 공유해주셔도 좋아요!

Gomding commented on src/test/java/roomescape/fake/FakeReservationDao.java:
```diff
@@ -0,0 +1,60 @@
+import roomescape.domain.Reservation;
+import roomescape.repository.ReservationRepository;
+
+public class FakeReservationDao implements ReservationRepository {
```
> 서비스 코드의 단위 테스트 대해서 추가로 질문드리고 싶은게 있습니다! 인기 테마 조회나 시간 이용 가능 여부 조회 같이 JOIN을 사용한 복잡한 쿼리를 사용하는 기능이 추가되니까 Fake 객체의 단점(구현이 복잡해짐, Fake 객체 자체의 오류 발생 가능)을 실감하고 있습니다. 따라서 서비스 단위 테스트를 위해 Fake 객체의 사용 대신 Mockito 라이브러리를 활용한 Mock의 도입을 고려하고 있습니다. 다만 너무 성급한 외부 라이브러리 도입이 아닐지 약간 우려가 되어서 찰리의 의견을 여쭙고 싶습니다. Mockito를 도입하기에 적절한 상황에 맞을까요? 혹은 Fake 객체의 사용만으로 충분할까요?

Mockito 사용시에는 그것을 100% 신뢰할 수 있을까요?
Fake 객체 사용과 동일한 문제를 안고있다고 생각해요 :)
(물론 그 사용은 간단해질 수 있지만요)

개인적인 의견이라
도입을 고려해보신다면 Mockito 를 사용해보고 비교해보셔도 좋아요!
Fake 와 비교해서 어떤 장단점이 있는지 직접 느껴보고 노랑의 기준을 만드는게 좋다고 생각해요 😃 
저랑 다른 결론을 내릴 수도 있으니까요

Gomding commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -0,0 +1,45 @@
+
+@RestController
+@RequestMapping("/reservations")
+public class ReservationController {
```
다음 단계 진행하면서 `API 레벨의 통합 테스트` 를 작성해보면 좋겠어요!

Gomding commented on src/main/java/roomescape/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,96 @@
+import org.springframework.web.method.annotation.MethodArgumentTypeMismatchException;
+
+@ControllerAdvice
+public class GlobalExceptionHandler {
```
노랑이 예측하지 못한 예외가 발생했을 때는 클라이언트에게 어떤 응답이 나가고있을까요? 🤔 
예측하지 못한 예외는 `GlobalExceptionHandler` 에서 별도 처리하지 않아도 될지 고민해보면 좋겠어요!

Gomding reviewed (approved):
안녕하세요 노랑!
이번 단계는 충분히 구현하신것 같아요 👍 
이만 머지하겠습니다
추가로 남긴 코멘트는 다음 단계에서 함께 고려해주세요!

궁금한 점 있으면 언제든 DM 이나 코멘트 남겨주세요
